Add support for WebKit2GTK+ in wxWebView

swt2c opened the pull request.

catalinr reviewed (commented):
Just some styling observations:
- where possible, local variables could be defined and initialized in the same statement (including static ones), and closer to the place where they are used, rather than at the beginning of a function;
- IMHO `goto`-s should be avoided, and most of them can be, by adding a short `else`, but probably all of them could be avoided by rearranging the code a bit.

swt2c commented:
Thanks for the review.

> where possible, local variables could be defined and initialized in the same statement (including static ones), and closer to the place where they are used, rather than at the beginning of a function;

I tend to agree with you.  A lot of this code was adopted from Epiphany which follows the C style of declaring all the variables at the beginning of the function.  I've left some of them in places to avoid exceeding 80-character lines (grrr, long gtk function names...).

> IMHO goto-s should be avoided, and most of them can be, by adding a short else, but probably all of them could be avoided by rearranging the code a bit.

I tend to think they're fine when used for errors/cleanup stuff, but sure, I can try to get rid of them.

swt2c commented on include/wx/gtk/webview_webkit.h:
```diff
@@ -134,6 +138,18 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewWebKit : public wxWebView
+    //Used for webkit2 extension
+    GDBusServer *m_dbus_server;
+    GDBusProxy *m_extension;
+#endif
```
Unfortunately, yes they do...they need to be accessed from C functions.

catalinr replied:
That crossed my mind too, and searching for them in the cpp and h files that you modified shows `m_findCount`, `m_dbus_server` and `m_extension`, each of them in one such situation. And that IMHO they would look better as getters. And at least the other 3 could co back to being protected.

swt2c replied:
Removed all added public members.

swt2c commented on include/wx/msw/setup0.h:
```diff
@@ -1444,9 +1455,9 @@
-// is used (although it will still save a few bytes probably).
+// that currently only wxMSW supports this so setting this to 0 doesn't change
+// much for non-MSW platforms (although it will still save a few bytes
+// probably).
```
Yep, I'm not sure where that came from.  I thought it was from regenerating bakefiles, but that doesn't seem to be the case.

swt2c replied:
Removed.

swt2c commented on src/gtk/webview_webkit2_extension.cpp:
```diff
@@ -0,0 +1,426 @@
+    void HasSelection(GVariant *parameters, GDBusMethodInvocation *invocation);
+
+    GDBusConnection *m_dbus_connection;
+    WebKitWebExtension *m_webkit_extension;
```
m_dbus_connection does.  m_webkit_extension appears that it does not, I'll correct that.

swt2c replied:
Updated to remove all public members.

swt2c commented on src/gtk/webview_webkit2_extension.cpp:
```diff
@@ -0,0 +1,426 @@
+    guint64 page_id;
+    g_variant_get(parameters, "(t)", &page_id);
+    WebKitWebPage *web_page;
+    web_page = webkit_web_extension_get_page(m_webkit_extension, page_id);
```
This is a case where I was trying to avoid exceeding 80 characters.  Obviously I didn't succeed in doing that throughout.  :(

catalinr replied:
Yes, that would be nice, and IMO a good attempt would also be something like
> 
    MyVariableType *var_pointer =
        long_gtk_function_name(lots, of, parameters);

swt2c replied:
Updated.

swt2c commented on src/gtk/webview_webkit2_extension.cpp:
```diff
@@ -0,0 +1,426 @@
+
+    if (!own_credentials)
+    {
+        own_credentials = g_credentials_new();
```
It probably would...this code came from Epiphany, I think.  I'll clean it up.

swt2c replied:
Updated.

swt2c commented on src/gtk/webview_webkit2_extension.cpp:
```diff
@@ -0,0 +1,426 @@
+    {
+        g_warning("Failed to connect to UI process: %s", error->message);
+        g_error_free(error);
+        return;
```
Not needed.  If error is non-NULL, then connection is NULL:
https://developer.gnome.org/gio/stable/GDBusConnection.html#g-dbus-connection-new-for-address-finish

swt2c commented on setup.h.in:
```diff
@@ -334,12 +334,18 @@
+#if defined(__WXGTK3__)
+#define wxUSE_WEBVIEW_WEBKIT2 0
+#else
+#define wxUSE_WEBVIEW_WEBKIT2 0
```
It looks like this file is auto-generated from include/wx/setup_inc.h, so I'm inclined to think it is fine as-is.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+
+    if (source)
+    {
+        wxString wxs = wxString(source, wxConvUTF8);
```
Yes, but I need to check whether source is non-NULL in order to free it, so I think I need to keep this as-is.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+            //case WEBKIT_POLICY_ERROR_FAILED:
+            //case WEBKIT_POLICY_ERROR_CANNOT_SHOW_MIME_TYPE:
+            //case WEBKIT_POLICY_ERROR_CANNOT_SHOW_URL:
+            //case WEBKIT_POLICY_ERROR_FRAME_LOAD_INTERRUPTED_BY_POLICY_CHANGE:
```
All of this code was inherited from src/gtk/webview_webkit.cpp, which I what I started with.  I just went ahead and removed all the commented out ones.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+    gtk_widget_set_sensitive(gtk_bin_get_child(GTK_BIN(m_widget)), enable);
+
+    //if (enable)
+    //    GTKFixSensitivity();
```
Same as the previous comment, was inherited code.  Removed.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+        free(source);
+        return wxs;
+    }
+    return wxString("");
```
Updated.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+    gboolean tozoom = webkit_settings_get_zoom_text_only(settings);
+
+    if (tozoom) return wxWEBVIEW_ZOOM_TYPE_TEXT;
+    else        return wxWEBVIEW_ZOOM_TYPE_LAYOUT;
```
Updated.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+            return wxString(text, wxConvUTF8);
+        }
+    }
+    return wxString("");
```
Updated.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+            return wxString(source, wxConvUTF8);
+        }
+    }
+    return wxString("");
```
Updated.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+            return wxString(text, wxConvUTF8);
+        }
+    }
+    return wxString("");
```
Updated.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+    m_findText = text;
+
+    //If the search string is empty then we clear any selection and highlight
+    if(text == "")
```
Updated.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1285 @@
+{
+    m_findCount = 0;
+    m_findFlags = 0;
+    m_findText = "";
```
Updated.

swt2c commented on src/gtk/webview_webkit2_extension.cpp:
```diff
@@ -0,0 +1,426 @@
+                                         GAsyncResult *result,
+                                         wxWebViewWebKitExtension *extension);
+
+static wxWebViewWebKitExtension *extension = NULL;
```
Updated.

swt2c commented on src/gtk/webview_webkit2_extension.cpp:
```diff
@@ -0,0 +1,426 @@
+    void GetSelectedText(GVariant *parameters, GDBusMethodInvocation *invocation);
+    void HasSelection(GVariant *parameters, GDBusMethodInvocation *invocation);
+
+    GDBusConnection *m_dbus_connection;
```
Updated.

swt2c commented:
Okay @catalinr, I think I have addressed all of your feedback.  Local variables should be defined and initialized in the same statement.  I removed all the gotos.  Also, I removed all the added public variables.  I was able to rework the code so they weren't needed.

Thanks again for the feedback.

sjlamerton commented:
Thanks for submitting this @swt2c and to @catalinr  for your review. I hope to have a more detailed look in the next day or so, but it looks great so far!

swt2c commented:
Just a few notes on the implementation:
- Everything is implemented and working AFAIK except for `ClearHistory()` and `EnableHistory()`.  Unfortunately the BackForwardList is read-only in WebKit2, so at the moment it doesn't seem possible to implement these functions in WebKit2.
- I used the same header file as the WebKit1 implementation (include/wx/gtk/webview_webkit.h) as there were not many changes required to the header.  There is a separate cpp file for the implementation, however: src/gtk/webview_webkit2.cpp.
- The WebKit2 extension is in src/gtk/webview_webkit2_extension.cpp.  The extension runs in a separate process and is required for accessing some of the WebKit2 API (namely DOM functionality).  I used DBus for IPC with the extension.  I considered using wx IPC, but I wasn't exactly sure how to integrate the wx event loop with the extension's event loop, and there were existing extension examples that used DBus.  Thus, the extension doesn't actually link to any wx libraries - it only uses the wx/defs.h header just to get the WXEXPORT definition.  I wasn't exactly sure of the best way to integrate the extension with the wx build system, so I ended up using the "plugin" mechanism which seems to work fine.
- In the cases where there are calls made to the secondary process, we do technically block until the response is received.  However, the secondary process seems to respond rather quickly, and I haven't noticed any perceptible delay in the UI.

sjlamerton commented:
> Everything is implemented and working AFAIK except for ClearHistory() and EnableHistory(). Unfortunately the BackForwardList is read-only in WebKit2, so at the moment it doesn't seem possible to implement these functions in WebKit2.

It looks like this is [by design](https://bugs.webkit.org/show_bug.cgi?id=147525) so leaving them out is fine.

> I used the same header file as the WebKit1 implementation (include/wx/gtk/webview_webkit.h) as there were not many changes required to the header. There is a separate cpp file for the implementation, however: src/gtk/webview_webkit2.cpp.

It would be good to add a note in the header that it is shared by both implementations.

> I wasn't exactly sure of the best way to integrate the extension with the wx build system, so I ended up using the "plugin" mechanism which seems to work fine.

@vadz Do you have an opinion on this?

Will take a more detailed look through the code later today.

swt2c commented:
> > I used the same header file as the WebKit1 implementation (include/wx/gtk/webview_webkit.h) as there were not many changes required to the header. There is a separate cpp file for the implementation, however: src/gtk/webview_webkit2.cpp.

> It would be good to add a note in the header that it is shared by both implementations.

Done.  I also added a bit of missing cleanup stuff to the destructor.

sjlamerton commented on configure.in:
```diff
@@ -7023,22 +7023,35 @@ fi
         if test "$wxUSE_GTK" = 1; then
             webkitgtk=webkit-1.0
             if test "$WXGTK3" = 1; then
                 webkitgtk=webkitgtk-3.0
```
I think you need to move this down into the `if test "$USE_WEBVIEW_WEBKIT2" = 0; then` section with the appropriate GTK3 check.

swt2c replied:
Unless I'm missing something, this won't change the functionality as that variable isn't used in the webkit2 case, but if you think it makes more logical sense to move it down there, OK.

swt2c replied:
Moved as suggested.

sjlamerton commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -0,0 +1,1265 @@
+
+void wxWebViewWebKit::EnableContextMenu(bool enable)
+{
+#if !WEBKIT_CHECK_VERSION(1, 10, 0) //If we are using an older version
```
Do we still need this check?

swt2c replied:
Probably not.  :-)  Left over from the wk1 implementation.

swt2c replied:
Removed.

sjlamerton reviewed (commented):
Mostly looks good to me, just a couple of minor points.

swt2c committed:
Add support for WebKit2GTK+ in wxWebView

sjlamerton commented:
Thanks for working on this!
